crates/kitsune_p2p/fetch/src/pool.rs

```diff
@@ -546,4 +653,193 @@ mod tests {
         assert_eq!(q.iter_mut(&cfg).count(), 3);
     }
+
+    #[tokio::test(start_paused = true)]
+    async fn state_iter_sees_all_items() {
+        let cfg = Config(1, 10);
+        let num_items = 2 * NUM_ITEMS_PER_POLL; // Must be great than NUM_ITEMS_PER_POLL
+
+        let mut q = {
+            let mut queue = vec![];
+            for i in 0..(num_items) {
+                queue.push((
+                    key_op(i as u8),
+                    item(&cfg, sources([(i % 100) as u8]), ctx(1)),
+                ))
+            }
+
+            State {
+                queue: queue.into_iter().collect(),
+            }
+        };
+
+        // None fetched initially, should see all items
+        assert_eq!(num_items, q.iter_mut(&cfg).count());
+
+        // Everything seen, no time elapsed
+        assert_eq!(0, q.iter_mut(&cfg).count());
+
+        // Move time forwards so everything will be ready to retry
+        tokio::time::advance(Duration::from_secs(30)).await;
+
+        assert_eq!(num_items, q.iter_mut(&cfg).count());
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn state_iter_uses_all_sources() {
+        let cfg = Config(1, 10);
+        let num_items = 10;
+
+        let mut q = {
+            let mut queue = vec![];
+            for i in 0..num_items {
+                queue.push((
+                    key_op(i as u8),
+                    // Give each item a different set of sources
+                    item(
+                        &cfg,
+                        sources((i * num_items) as u8..(i * num_items + num_items) as u8),
+                        ctx(1),
+                    ),
+                ))
+            }
+
+            State {
+                queue: queue.into_iter().collect(),
+            }
+        };
+
+        let mut seen_sources = HashSet::new();
+        for _ in 0..num_items {
+            q.iter_mut(&cfg)
+                .map(|item| match item.2 {
+                    FetchSource::Agent(a) => a.0.clone(),
+                })
+                .for_each(|source| {
+                    seen_sources.insert(source);
+                });
+
+            // Move time forwards so everything will be ready to retry
+            tokio::time::advance(Duration::from_secs(30)).await;
+        }
+
+        assert_eq!(num_items * num_items, seen_sources.len());
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn remove_fetch_item() {
+        let cfg = Config(1, 10);
+        let mut q = {
+            let queue = [(key_op(1), item(&cfg, sources([1]), ctx(1)))];
+
+            let queue = queue.into_iter().collect();
+            State { queue }
+        };
+
+        assert_eq!(1, q.iter_mut(&cfg).count());
+        q.remove(&key_op(1));
+
+        // Move time forwards to be able to retry the item
+        tokio::time::advance(Duration::from_secs(30)).await;
+
+        assert_eq!(0, q.iter_mut(&cfg).count());
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn fetch_pool() {
```

Comment:
We weren't testing through the main interface that's exposed from the crate. This test is meant to be representative of how the interface is actually used in Kitsune.